engine/access/rest/websockets/data_providers/events_provider.go

```diff
@@ -0,0 +1,175 @@
+package data_providers
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine/access/rest/common/parser"
+	"github.com/onflow/flow-go/engine/access/rest/http/request"
+	"github.com/onflow/flow-go/engine/access/rest/util"
+	"github.com/onflow/flow-go/engine/access/rest/websockets/models"
+	"github.com/onflow/flow-go/engine/access/state_stream"
+	"github.com/onflow/flow-go/engine/access/subscription"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/counters"
+)
+
+// EventsArguments contains the arguments required for subscribing to events
+type EventsArguments struct {
+	StartBlockID     flow.Identifier          // ID of the block to start subscription from
+	StartBlockHeight uint64                   // Height of the block to start subscription from
+	Filter           state_stream.EventFilter // Filter applied to events for a given subscription
+}
+
+// EventsDataProvider is responsible for providing events
+type EventsDataProvider struct {
+	*baseDataProvider
+
+	logger         zerolog.Logger
+	stateStreamApi state_stream.API
+}
+
+var _ DataProvider = (*EventsDataProvider)(nil)
+
+// NewEventsDataProvider creates a new instance of EventsDataProvider.
+func NewEventsDataProvider(
+	ctx context.Context,
+	logger zerolog.Logger,
+	stateStreamApi state_stream.API,
+	chain flow.Chain,
+	eventFilterConfig state_stream.EventFilterConfig,
+	topic string,
+	arguments models.Arguments,
+	send chan<- interface{},
```

Comment:
nit: keep the argument order the same as the other methods
```suggestion
	ctx context.Context,
	logger zerolog.Logger,
	stateStreamApi state_stream.API,
	topic string,
	arguments models.Arguments,
	send chan<- interface{},
	chain flow.Chain,
	eventFilterConfig state_stream.EventFilterConfig,
```